presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -1659,13 +1666,19 @@ private void undoAlterPartitionOperations()
 
         private void undoUpdateStatisticsOperations()
         {
+            ImmutableList.Builder<CompletableFuture<?>> undoUpdateFutures = ImmutableList.builder();
             for (UpdateStatisticsOperation operation : updateStatisticsOperations) {
-                try {
-                    operation.undo(delegate);
-                }
-                catch (Throwable throwable) {
-                    logCleanupFailure(throwable, "failed to rollback: %s", operation.getDescription());
-                }
+                undoUpdateFutures.add(CompletableFuture.runAsync(() -> {
```

Comment:
I think now there can be a race condition if one of the update futures fails, but others continue. In such case update and undo operations would be executed at the same time

Reply:
This should not happen after the latest changes.